cloudinit/sources/DataSourceVMware.py

```diff
@@ -0,0 +1,814 @@
+        LOG.info("got host-info: %s", host_info)
+
+        # Reflect any possible local IPv4 or IPv6 addresses in the guest
+        # info.
+        advertise_local_ip_addrs(host_info)
+
+        # Ensure the metadata gets updated with information about the
+        # host, including the network interfaces, default IP addresses,
+        # etc.
+        self.metadata = util.mergemanydict([self.metadata, host_info])
+
+        # Persist the instance data for versions of cloud-init that support
+        # doing so. This occurs here rather than in the get_data call in
+        # order to ensure that the network interfaces are up and can be
+        # persisted with the metadata.
+        self.persist_instance_data()
+
+    def _get_subplatform(self):
+        get_key_name_fn = None
+        if self.data_access_method == DATA_ACCESS_METHOD_ENVVAR:
+            get_key_name_fn = get_guestinfo_envvar_key_name
+        elif self.data_access_method == DATA_ACCESS_METHOD_GUESTINFO:
+            get_key_name_fn = get_guestinfo_key_name
+        else:
+            return sources.METADATA_UNKNOWN
+
+        return "%s (%s)" % (
+            self.data_access_method,
+            get_key_name_fn("metadata"),
+        )
+
+    @property
+    def network_config(self):
+        if "network" in self.metadata:
+            LOG.debug("using metadata network config")
+        else:
+            LOG.debug("using fallback network config")
+            self.metadata["network"] = {
+                "config": self.distro.generate_fallback_config(),
+            }
+        return self.metadata["network"]["config"]
+
+    def get_instance_id(self):
+        # Pull the instance ID out of the metadata if present. Otherwise
+        # read the file /sys/class/dmi/id/product_uuid for the instance ID.
+        if self.metadata and "instance-id" in self.metadata:
+            return self.metadata["instance-id"]
+        with open(PRODUC_UUID_FILE_PATH, "r") as id_file:
+            self.metadata["instance-id"] = str(id_file.read()).rstrip().lower()
+            return self.metadata["instance-id"]
+
+    def get_public_ssh_keys(self):
+        for key_name in (
+            "public-keys-data",
+            "public_keys_data",
+            "public-keys",
+            "public_keys",
+        ):
+            if key_name in self.metadata:
+                return sources.normalize_pubkey_data(self.metadata[key_name])
+        return []
+
+    def redact_keys(self):
+        # Determine if there are any keys to redact.
+        keys_to_redact = None
+        if REDACT in self.metadata:
+            keys_to_redact = self.metadata[REDACT]
+        elif CLEANUP_GUESTINFO in self.metadata:
+            # This is for backwards compatibility.
+            keys_to_redact = self.metadata[CLEANUP_GUESTINFO]
+
+        if self.data_access_method == DATA_ACCESS_METHOD_GUESTINFO:
+            guestinfo_redact_keys(keys_to_redact, self.vmware_rpctool)
+
+
+def decode(key, enc_type, data):
```

Comment:
This really has a couple entry points that could be satisfied by the existing util.b64d, decode_binary and util.decomp_gzip. Please use these common util functions where possible to meet the needs of your decode function.

Reply:
Sure. I did this in part for the logging about the keys being used, but as I adjust logging maybe this is no longer an issue and I can switch.